src/lib.rs

```diff
@@ -591,30 +592,84 @@ impl<T, C: cfg::Config> Shard<T, C> {
         let shared = &self.shared[page_index];
         shared.remove(addr, C::unpack_gen(idx), shared.free_list())
     }
+}
 
-    /// Remove an item on the shard's local thread.
-    fn take_local(&self, idx: usize) -> Option<T> {
+impl<T, C> Shard<T, C>
+where
+    T: clear::Clear + Default,
+    C: cfg::Config,
+{
+    pub(crate) fn get_initialized_slot(&self) -> Option<usize> {
+        for (page_idx, page) in self.shared.iter().enumerate() {
+            let local = self.local(page_idx);
+
+            test_println!("-> page {}; {:?}; {:?}", page_idx, local, page);
+
+            if let offset @ Some(_) = page.get_initialized_slot(local) {
+                return offset;
+            }
+        }
+
+        None
+    }
+
+    pub(crate) fn new(tid: usize) -> Self {
+        let mut total_sz = 0;
+        let shared = (0..C::MAX_PAGES)
+            .map(|page_num| {
+                let sz = C::page_size(page_num);
+                let prev_sz = total_sz;
+                total_sz += sz;
+                page::Shared::new(sz, prev_sz)
+            })
+            .collect();
+        let local = (0..C::MAX_PAGES).map(|_| page::Local::new()).collect();
+        Self { tid, local, shared }
+    }
+
+    pub(crate) fn clear_local(&self, idx: usize) -> bool {
         debug_assert_eq!(Tid::<C>::from_packed(idx).as_usize(), self.tid);
         let (addr, page_index) = page::indices::<C>(idx);
 
-        test_println!("-> remove_local {:?}", addr);
+        if page_index > self.shared.len() {
+            return false;
+        }
 
-        self.shared
-            .get(page_index)?
-            .take(addr, C::unpack_gen(idx), self.local(page_index))
+        // TODO (bhargav): god this is icky
+        if let Some(true) =
+            self.shared[page_index].clear(addr, C::unpack_gen(idx), self.local(page_index))
+        {
+            true
+        } else {
+            false
+        }
```

Comment:
i think you could write
```suggestion
        self.shared[page_index]
            .clear(addr, C::unpack_gen(idx), self.local(page_index))
            .unwrap_or(false)
```
instead?

Reply:
See that's what I get for programming instead of going to sleep.